### ISSUE TITLE
opt: transform some of the expressions into equivalent CASE expressions

### ISSUE THREAD

**Issue by yuzefovich** [cockroachdb/cockroach#77799]
In the vectorized execution engine we use the CASE operator to support several expressions. We do so during the vectorized planning, but it seems more appropriate to do the transformation during the normalization stage of the optimization.

I believe that the execution in the row-by-row engine will not suffer performance-wise, so we'll be able to remove the implementations of `Eval` methods for the corresponding expressions.

In particular, the following constructs are implemented in the vectorized planning:
- `WHERE a OR b`
  - into `WHERE CASE WHEN a THEN true WHEN b THEN true ELSE false END`
- `COALESCE(a, b, ...)`
  - into `CASE WHEN a IS DISTINCT FROM NULL THEN a WHEN b IS DISTINCT FROM NULL THEN b ... END`
- `IF(a, b, c)`
  - into `CASE WHEN a THEN b ELSE c END`
- `NULLIF(a, b)`
  - into `CASE WHEN a = b THEN NULL ELSE a END`.

**Comment by michae2** [comment]
One problem with doing this during normalization is that it will interfere with exploration rules that match on OR such as SplitDisjunction. It also seems a bit confusing to someone reading EXPLAIN output to see their OR predicate rewritten in this way... it seems like the vectorized engine is the only consumer that wants this rewrite? Is there a benefit to moving this rewrite out of the vectorized engine other than it seeming more appropriate?

**Comment by yuzefovich** [comment]
Good point - I don't think I considered broader impact of this change on other rules that might be performed by the optimizer. Indeed, the main motivation was that it seemed wrong to have normatlization-like rule implemented in the vectorized engine, so I filed this issue. We probably should close it as a bad idea.

**Comment by BryanWangAtCRL** [comment]
Alright, it's good that we are all in agreement; I'll omit the code for OR and just make changes related to COALESCE. Thanks for taking a look!

**Comment by BryanWangAtCRL** [comment]
Actually, I think there might be some hidden issues with the COALESCE implementation (please correct me if I'm wrong) because you query the same thing twice now.

A bit gimmicky counter example that I ran on postgresql:

CREATE SEQUENCE seq;

 SELECT COALESCE(nextval('seq'), -1);

Outputs: 1

==============================

Versus the translated version:

CREATE SEQUENCE seq;
  
 SELECT CASE
    WHEN nextval('seq') IS DISTINCT FROM NULL THEN nextval('seq')   -- two separate calls
    WHEN -1 IS DISTINCT FROM NULL THEN -1
    ELSE NULL
  END;

Outputs: 2


Which is a bit of a problem. Basically, it doesn't work for volatile stuff. I'll also try running it on cockroach in a bit.


There's another failure that the test cases catch. I'm still trying to understand it fully, so here's an explanation from Claude.

===========================
Correlated subquery argument:

  CREATE TABLE t (a INT);
  CREATE TABLE u (x INT, y INT);
  
  SELECT COALESCE((SELECT y FROM u WHERE x = t.a), 0) FROM t;
  * master: plans normally
  *  with rewrite: assertion failure
   * "left-join RelExpr children have intersecting columns"

Why: the correlated subquery (SELECT y FROM u WHERE x \= t.a) gets duplicated into the WHEN condition and the THEN value. Decorrelation hoists each occurrence into its own apply-join, and both joins claim the same subquery output column CheckExpr rejects the malformed plan.

===========================

I'll try to investigate it a bit more since it seems like other similar transformations have already been implemented. Perhaps the test case fail can be mitigated, but it does seem like it'll change results regardless.

**Comment by BryanWangAtCRL** [comment]
PostgreSQL 7.1.3 Documentation does mention this error with NULLIF and COALESCE. But the current documentation makes no mention.

"Tip: COALESCE and NULLIF are just shorthand for CASE expressions. They are actually converted into CASE expressions at a very early stage of processing, and subsequent processing thinks it is dealing with CASE. Thus an incorrect COALESCE or NULLIF usage may draw an error message that refers to CASE."

I'll first try to investigate the error the testcases catch.

**Comment by michae2** [comment]
Interesting! Looks like the existing rewrite in the vectorized engine also causes volatile expressions to be evaluated twice:

```sql
CREATE SEQUENCE s;
CREATE TABLE t (a INT PRIMARY KEY);
INSERT INTO t VALUES (1);

-- row-by-row engine:
SET vectorize = off;
SELECT COALESCE(nextval('s'), 0) FROM t;
SELECT currval('s');
-- 1

-- vectorized engine:
ALTER SEQUENCE s RESTART WITH 1;
SET vectorize = on;
SELECT COALESCE(nextval('s'), 0) FROM t;
SELECT currval('s');
-- 2
```

The [PG docs](https://www.postgresql.org/docs/current/functions-conditional.html#FUNCTIONS-COALESCE-NVL-IFNULL) on COALESCE mention that it evaluates arguments lazily, short-circuiting after the first non-null value. But they don't mention how many times each argument is evaluated. In practice it seems to be once. The [SQL Server docs](https://learn.microsoft.com/en-us/sql/t-sql/language-elements/coalesce-transact-sql?view=sql-server-ver17) mention a rewrite into CASE, and that expressions are evaluated multiple times. It seems like both behaviors are compliant with the SQL spec. But it would be nice to be consistent with PG (not to mention internally consistent!).

I'm not sure I agree that rewriting COALESCE to CASE in optbuild or as a normalization rule is an improvement over the current behavior. We already have some normalization rules that operate directly on COALESCE, it would be nice to keep those. And EvalCoalesceExpr in the row-by-row engine more closely matches PG behavior than the rewrite to CASE in the vectorized engine. If anything, it seems we should make the vectorized engine match the row-by-row engine, rather than lifting the vectorized engine's rewrite up. @yuzefovich do you have any opinion?

**Comment by michae2** [comment]
(Also, subqueries are always evaluated eagerly for a different reason, even in COALESCE, which is captured in https://github.com/cockroachdb/cockroach/issues/20298. Best to stick to volatile scalar expressions for the purposes of this discussion.)

**Comment by yuzefovich** [comment]
Nice finds! I agree that we should unify both engines and match postgres. We probably want to reject the transformation to CASE expression in `execplan.go` when it's invalid since adjusting the behavior to match these one-time-only semantics seems tricky and fragile.

A related problem seems to be with NULLIF expression:
```sql
CREATE SEQUENCE s;
CREATE TABLE t (a INT PRIMARY KEY);
INSERT INTO t VALUES (1);
-- both engines agree due
-- SET vectorize = off;
SELECT NULLIF(nextval('s'), nextval('s')) FROM t;
-- PG returns 1, CRDB returns 3
SELECT currval('s');
-- PG returns 2, CRDB returns 3
```
This appears to be because of some optimizer transformation:
```sql
EXPLAIN (VERBOSE) SELECT NULLIF(nextval('s'), nextval('s')) FROM t;                                                  
--  • render
--  │ columns: ("nullif")
--  │ render nullif: CASE nextval('s') WHEN nextval('s') THEN CAST(NULL AS INT8) ELSE nextval('s') END
```

**Comment by BryanWangAtCRL** [comment]
I'm currently trying to understand how to that.

But before that, I just want to point out that these commands crash master right based on the case 2 that I mentioned above.

CREATE TABLE a (k INT PRIMARY KEY, i INT);
CREATE TABLE xy (x INT PRIMARY KEY, y INT);
SELECT NULLIF((SELECT y FROM xy WHERE x = a.i), 0) FROM a;  


Stack Trace if you are interested:

ERROR: internal error: top-level relational expression cannot have outer columns: (8)
SQLSTATE: XX000
DETAIL: stack trace:
pkg/sql/opt/xform/optimizer.go:282: Optimize()
pkg/sql/plan_opt.go:1097: buildExecMemo()
pkg/sql/plan_opt.go:364: makeOptimizerPlanInternal()
pkg/sql/plan_opt.go:344: makeOptimizerPlan()
pkg/sql/conn_executor_exec.go:3562: makeExecPlan()
pkg/sql/conn_executor_exec.go:3111: dispatchToExecutionEngine()
pkg/sql/conn_executor_exec.go:1144: execStmtInOpenState()
pkg/sql/conn_executor_exec.go:185: func2()
pkg/sql/conn_executor_exec.go:4915: execWithProfiling()
pkg/sql/conn_executor_exec.go:184: execStmt()
pkg/sql/conn_executor.go:2671: func1()
pkg/sql/conn_executor.go:2676: execCmd()
pkg/sql/conn_executor.go:2553: run()
pkg/sql/conn_executor.go:1225: ServeConn()
pkg/sql/pgwire/conn.go:254: processCommands()
pkg/sql/pgwire/server.go:1337: func4()
src/runtime/asm_arm64.s:1447: goexit()

**Comment by michae2** [comment]
@BryanWangAtCRL and I did a little debugging together and figured out that [this code](https://github.com/cockroachdb/cockroach/blob/32b032d7df26f777ec5218e511f0bca967641128/pkg/sql/opt/optbuilder/scalar.go#L400) in optbuilder is duplicating the subquery with the same column IDs, which causes problems for the decorrelation normalization rules later on.

I think this is another instance of the "duplicate subqueries" problem @DrewKimball is working on for https://github.com/cockroachdb/cockroach/issues/116022 / https://github.com/cockroachdb/cockroach/issues/155681 and we should set it aside for now.
